### PR TITLE
Optional evaluations and better errors. Closes broadinstitute/cromwell#1830

### PR DESCRIPTION
**Pull request by cjllanwarne**

### CODE REVIEW

File: src/main/scala/wdl4s/values/WdlValue.scala [hunk 1]
```diff
@@ -11,10 +11,10 @@ import scala.util.{Failure, Try}
 trait WdlValue {
   val wdlType: WdlType
   def invalid(operation: String) = Failure(new WdlExpressionException(s"Cannot perform operation: $operation"))
-  def emptyValue(value: WdlOptionalValue) = Failure(VariableNotFoundException(value.toString))
-  def evaluateIfDefined[A <: WdlValue](optionalValue: WdlOptionalValue, operation: WdlValue => Try[A]): Try[A] = optionalValue match {
+  def emptyValueFailure(operationName: String) = Failure(new UnsupportedOperationException(s"Sorry! Operation $operationName is not supported on empty optional values. You might resolve this using select_first(optional, default) to guarantee that you have a filled value."))
```
Comment (cjllanwarne) [inline src/main/scala/wdl4s/values/WdlValue.scala]: Pinging @katevoss for critique!

Comment (cjllanwarne) [inline src/main/scala/wdl4s/values/WdlValue.scala]: Maybe should be `select_first([optional, default])`

Comment (mcovarr) [inline src/main/scala/wdl4s/values/WdlValue.scala]: filled?  like a doughnut?

**Comment by mcovarr** [comment]
:+1: monads are like burritos, options are like doughnuts.

**Comment by cjllanwarne** [comment]
@katevoss that `+` is from a test case which tries something invalid with `+`. Whatever `operationName` is  gets filled in into the error message as appropriate.

**Comment by Horneth** [comment]
👍

**Comment by cjllanwarne** [comment]
Will return later!

**Pull request closed without merging**